src/v/container/include/container/fragmented_hash_map.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include <absl/hash/hash.h>
+#include <ankerl/unordered_dense.h>
+#include <container/fragmented_vector.h>
+
+#include <type_traits>
+
+namespace detail {
+
+template<typename T>
+concept has_absl_hash = requires(T val) {
+    { AbslHashValue(std::declval<absl::HashState>(), val) };
+};
+
+/// Wrapper around absl::Hash that disables the extra hash mixing in
+/// unordered_dense
+template<typename T>
+struct avalanching_absl_hash {
+    // absl always hash mixes itself so no need to do it again
+    using is_avalanching = void;
+
+    auto operator()(const T& x) const noexcept -> uint64_t {
+        return absl::Hash<T>()(x);
+    }
+};
+
+} // namespace detail
+
+/**
+ * @brief A hash map that uses a fragmented vector as the underlying storage.
+ *
+ * Use when the hash map is expected to have a large number of elements (e.g.:
+ * scales with partitions or topics). Performance wise it's equal to the abseil
+ * hashmaps.
+ *
+ * NB: References and iterators are not stable across insertions and deletions.
+ *
+ * Both std::hash and abseil's AbslHashValue are supported. We dispatch to the
+ * later if available. Given AbslHashValue also supports std::hash we could also
+ * unconditionally dispatch to it. However, it's hash mixing seems more
+ * extensive (and hence less performant) so we only do that when needed.
```

Comment:
I didn't follow this: when you say "its hash mixing ..." is that absl hash or std hash?

Reply:
Meant absl hash (std hash doesn't do any hash mixing). Have clarified that.